File: src/Core/Rounding.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace NinjaCalc.Core {

    /// <summary>
    /// Static class to help with the rounding of calculator variables, especially
    /// with significant figure-based rounding.
    /// </summary>
    public static class Rounding {

        /// <summary>
        /// Round a number to the specified number of significant figures.
        /// </summary>
        /// <param name="numberToRound">The number you wish to round.</param>
        /// <param name="digits">The number of significant figures you wish to round the number to.</param>
        /// <returns>The rounded number.</returns>
        public static double RoundToSignificantDigits(this double numberToRound, int digits) {
            if (numberToRound == 0)
                return 0;

            double scale = Math.Pow(10, Math.Floor(Math.Log10(Math.Abs(numberToRound))) + 1);
            return scale * Math.Round(numberToRound / scale, digits);
        }

    }
}

File: src/Core/Rounding.java
package Core;

/**
 * Static class to help with the rounding of calculator variables, especially
 * with significant figure-based rounding.
 * @author gbmhunter
 * @since 2015-11-02
 */
public class Rounding {

    /**
     * Round a number to the specified number of significant figures.
     * @param numberToRound     The number you wish to round.
     * @param digits            The number of significant figures you wish to round the number to.
     * @return                  The rounded number.
     */
    public static double RoundToSignificantDigits(double numberToRound, int digits) {
        //System.out.println("RoundToSignificantDigits() called.");

        if(numberToRound == 0) {
            return 0;
        }

        final double d = Math.ceil(Math.log10(numberToRound < 0 ? -numberToRound: numberToRound));
        final int power = digits - (int) d;

        final double magnitude = Math.pow(10, power);
        final long shifted = Math.round(numberToRound*magnitude);
        return shifted/magnitude;
    }

}

